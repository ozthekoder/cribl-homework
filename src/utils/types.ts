export interface Log {
  Time: string;
  Event: string;
}
