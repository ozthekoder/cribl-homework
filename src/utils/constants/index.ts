export const LOGS_URL =
  "https://s3.amazonaws.com/io.cribl.c021.takehome/cribl.log";
