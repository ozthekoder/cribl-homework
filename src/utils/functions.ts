export function getJsonLineCount(obj: any): number {
  const formattedJson = JSON.stringify(obj, null, 2);
  return formattedJson.split("\n").length;
}


