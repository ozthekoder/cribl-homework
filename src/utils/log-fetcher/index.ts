import { Log } from "../types";

export interface LogFetcherProps {
  onLogs: (logs: Log[]) => void; // Send logs in batch
  batchSize?: number; // Configurable batch size
  chunkSize?: number; // Configurable chunk size (default: 16KB)
}

export class LogFetcher {
  public onLogs: (logs: Log[]) => void;
  private batchSize: number;
  private chunkSize: number;
  private logBatch: Log[] = [];

  public constructor(props: LogFetcherProps) {
    this.onLogs = props.onLogs;
    this.batchSize = props.batchSize ?? 10; // Default batch size
    this.chunkSize = props.chunkSize ?? 16 * 1024; // Default 16KB
  }

  public setOnLogs(onLogs: (logs: Log[]) => void) {
    this.onLogs = onLogs;
  }

  public async fetch(url: string) {
    const response = await fetch(url);

    if (!response.ok) {
      throw new Error(`Failed to fetch: ${response.statusText}`);
    } else if (response.body) {
      const reader = response.body.getReader();
      const decoder = new TextDecoder("utf-8");

      let leftover = ""; // Store partial lines between chunks
      let buffer = new Uint8Array(this.chunkSize); // Buffer to store limited chunk data
      let bufferOffset = 0; // Tracks how much of the buffer is filled

      while (true) {
        const { done, value } = await reader.read();
        if (done) break;

        let start = 0;
        while (start < value.byteLength) {
          // Determine how much we can take from `value`
          let bytesToCopy = Math.min(
            this.chunkSize - bufferOffset,
            value.byteLength - start,
          );

          // Copy chunk data into buffer
          buffer.set(value.subarray(start, start + bytesToCopy), bufferOffset);
          bufferOffset += bytesToCopy;
          start += bytesToCopy;

          // If buffer is full, process the chunk
          if (bufferOffset === this.chunkSize) {
            let textChunk = decoder.decode(buffer, { stream: true });

            let combinedText = leftover + textChunk;
            let linesReceived = combinedText.split("\n");

            leftover = linesReceived.pop() || ""; // Retain last partial line
            this.processLines(linesReceived);

            bufferOffset = 0; // Reset buffer for next chunk
          }
        }
      }

      // Process remaining data in the buffer
      if (bufferOffset > 0) {
        let textChunk = decoder.decode(buffer.subarray(0, bufferOffset), {
          stream: true,
        });
        let combinedText = leftover + textChunk;
        let linesReceived = combinedText.split("\n");

        leftover = linesReceived.pop() || "";
        this.processLines(linesReceived);
      }

      // Send any final leftover lines
      if (leftover) {
        this.logBatch.push(LogFetcher.processLine(leftover));
      }

      this.flushBatch(); // Ensure last batch is sent
    }
  }

  private processLines(lines: string[]) {
    for (let line of lines) {
      this.logBatch.push(LogFetcher.processLine(line));

      // Send batch if limit is reached
      if (this.logBatch.length >= this.batchSize) {
        this.flushBatch();
      }
    }
  }

  private flushBatch() {
    if (this.logBatch.length > 0) {
      this.onLogs([...this.logBatch]); // Send batch
      this.logBatch = []; // Clear batch
    }
  }

  public static processLine(line: string): Log {
    try {
      const json = JSON.parse(line);
      return { Time: new Date(json._time).toISOString(), Event: line };
    } catch (e) {
      return { Time: new Date().toISOString(), Event: line };
    }
  }
}
