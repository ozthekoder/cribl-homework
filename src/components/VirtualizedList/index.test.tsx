import { render, fireEvent, screen, act } from "@testing-library/react";
import { VirtualizedList } from ".";
import { Log } from "../../utils/types";
import "@testing-library/jest-dom";

beforeAll(() => {
  jest.doMock("../../utils/functions", () => ({
    getJsonLineCount: jest.fn(() => 10),
  }));
});

const logs: Log[] = Array.from({ length: 100 }, (_, i) => ({
  Time: `2024-03-01 12:${String(i).padStart(2, "0")}:00`,
  Event: JSON.stringify({ message: `Log ${i}` }),
}));

beforeEach(() => {
  jest.clearAllMocks();
});

const RowComponent = ({
  index,
  data,
  expanded,
  rowHeight,
}: {
  index: number;
  data: Log;
  expanded: boolean;
  rowHeight: number;
}) => (
  <div
    data-testid={`row-${index}`}
    data-index={index}
    data-expanded={expanded}
    style={{ height: rowHeight }}
  >
    {data.Time}
  </div>
);

describe("VirtualizedList Component", () => {
  it("should render only the visible rows", async () => {
    render(
      <VirtualizedList
        height={500}
        itemCount={logs.length}
        itemSize={32}
        itemData={logs}
      >
        {RowComponent}
      </VirtualizedList>,
    );

    const renderedRows = await screen.findAllByTestId(/^row-/);
    expect(renderedRows.length).toBeGreaterThan(0);
    expect(renderedRows.length).toBeLessThan(logs.length);
  });

  it("should update visible rows when scrolled", async () => {
    const { getByTestId } = render(
      <VirtualizedList
        height={320}
        itemCount={logs.length}
        itemSize={32}
        itemData={logs}
      >
        {RowComponent}
      </VirtualizedList>,
    );

    const container = getByTestId("virtualized-container");

    let initialRows = screen
      .getAllByTestId(/^row-/)
      .map((row) => row.dataset.index);
    expect(initialRows.length).toBeGreaterThan(0);

    fireEvent.scroll(container, { target: { scrollTop: 64 } });

    let updatedRows = screen
      .getAllByTestId(/^row-/)
      .map((row) => row.dataset.index);

    expect(updatedRows).not.toEqual(initialRows);
    expect(updatedRows.length).toBe(initialRows.length);
  });

  it("should increase height when row is expanded", async () => {
    render(
      <VirtualizedList
        height={500}
        itemCount={logs.length}
        itemSize={32}
        itemData={logs}
      >
        {RowComponent}
      </VirtualizedList>,
    );

    let row = await screen.findByTestId("row-1");

    const initialHeight = parseInt(row.style.height, 10);
    expect(initialHeight).toBe(32);

    fireEvent(
      row,
      new MouseEvent("click", {
        bubbles: true,
        cancelable: true,
      }),
    );

    row = await screen.findByTestId("row-1");

    const expandedHeight = parseInt(row.style.height, 10);
    expect(expandedHeight).toBeGreaterThan(32);
  });

  it("should collapse row when clicked again", async () => {
    render(
      <VirtualizedList
        height={500}
        itemCount={logs.length}
        itemSize={32}
        itemData={logs}
      >
        {RowComponent}
      </VirtualizedList>,
    );

    const row = await screen.findByTestId("row-0");

    fireEvent(
      row,
      new MouseEvent("click", {
        bubbles: true,
        cancelable: true,
      }),
    );

    fireEvent(
      row,
      new MouseEvent("click", {
        bubbles: true,
        cancelable: true,
      }),
    );

    expect(row).toHaveAttribute("data-expanded", "false");
    expect(parseInt(row.style.height, 10)).toBe(32);
  });

  it("should match snapshot", () => {
    const { container } = render(
      <VirtualizedList
        height={500}
        itemCount={logs.length}
        itemSize={32}
        itemData={logs}
      >
        {RowComponent}
      </VirtualizedList>,
    );

    expect(container.firstChild).toMatchSnapshot();
  });
});
