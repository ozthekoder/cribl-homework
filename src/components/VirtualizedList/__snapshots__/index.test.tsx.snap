// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`VirtualizedList Component should match snapshot 1`] = `
<div
  class="css-10b489u"
  data-testid="virtualized-container"
  style="height: 500px;"
>
  <div
    class="css-lg0a9e"
  >
    <div
      class="css-9db80r"
      data-testid="virtualized-rows"
    >
      <div
        data-expanded="false"
        data-index="0"
        data-testid="row-0"
        style="height: 32px;"
      >
        2024-03-01 12:00:00
      </div>
      <div
        data-expanded="false"
        data-index="1"
        data-testid="row-1"
        style="height: 32px;"
      >
        2024-03-01 12:01:00
      </div>
      <div
        data-expanded="false"
        data-index="2"
        data-testid="row-2"
        style="height: 32px;"
      >
        2024-03-01 12:02:00
      </div>
      <div
        data-expanded="false"
        data-index="3"
        data-testid="row-3"
        style="height: 32px;"
      >
        2024-03-01 12:03:00
      </div>
      <div
        data-expanded="false"
        data-index="4"
        data-testid="row-4"
        style="height: 32px;"
      >
        2024-03-01 12:04:00
      </div>
      <div
        data-expanded="false"
        data-index="5"
        data-testid="row-5"
        style="height: 32px;"
      >
        2024-03-01 12:05:00
      </div>
      <div
        data-expanded="false"
        data-index="6"
        data-testid="row-6"
        style="height: 32px;"
      >
        2024-03-01 12:06:00
      </div>
      <div
        data-expanded="false"
        data-index="7"
        data-testid="row-7"
        style="height: 32px;"
      >
        2024-03-01 12:07:00
      </div>
      <div
        data-expanded="false"
        data-index="8"
        data-testid="row-8"
        style="height: 32px;"
      >
        2024-03-01 12:08:00
      </div>
      <div
        data-expanded="false"
        data-index="9"
        data-testid="row-9"
        style="height: 32px;"
      >
        2024-03-01 12:09:00
      </div>
      <div
        data-expanded="false"
        data-index="10"
        data-testid="row-10"
        style="height: 32px;"
      >
        2024-03-01 12:10:00
      </div>
      <div
        data-expanded="false"
        data-index="11"
        data-testid="row-11"
        style="height: 32px;"
      >
        2024-03-01 12:11:00
      </div>
      <div
        data-expanded="false"
        data-index="12"
        data-testid="row-12"
        style="height: 32px;"
      >
        2024-03-01 12:12:00
      </div>
      <div
        data-expanded="false"
        data-index="13"
        data-testid="row-13"
        style="height: 32px;"
      >
        2024-03-01 12:13:00
      </div>
      <div
        data-expanded="false"
        data-index="14"
        data-testid="row-14"
        style="height: 32px;"
      >
        2024-03-01 12:14:00
      </div>
      <div
        data-expanded="false"
        data-index="15"
        data-testid="row-15"
        style="height: 32px;"
      >
        2024-03-01 12:15:00
      </div>
      <div
        data-expanded="false"
        data-index="16"
        data-testid="row-16"
        style="height: 32px;"
      >
        2024-03-01 12:16:00
      </div>
      <div
        data-expanded="false"
        data-index="17"
        data-testid="row-17"
        style="height: 32px;"
      >
        2024-03-01 12:17:00
      </div>
      <div
        data-expanded="false"
        data-index="18"
        data-testid="row-18"
        style="height: 32px;"
      >
        2024-03-01 12:18:00
      </div>
      <div
        data-expanded="false"
        data-index="19"
        data-testid="row-19"
        style="height: 32px;"
      >
        2024-03-01 12:19:00
      </div>
      <div
        data-expanded="false"
        data-index="20"
        data-testid="row-20"
        style="height: 32px;"
      >
        2024-03-01 12:20:00
      </div>
    </div>
  </div>
</div>
`;
